### PR TITLE
Initial commit of input ktable ttl example

### PR DESCRIPTION
**Pull request by sarwarbhuiyan**
### Description

Kafka tutorial showing how to periodically clean up a KTable based on a TTL. This only works for KTables instantiated from kafka topics and not in the middle of topologies but it is still a useful example to demonstrate.

### Staging Docs

http://kafka-tutorials-staging.s3-website-us-west-2.amazonaws.com/ktable-ttl
http://kafka-tutorials-staging.s3-website-us-west-2.amazonaws.com/ktable-ttl/schedule-ktable-ttl/kstreams.html

### CODE REVIEW

**Comment by bbejeck** [comment]
Thanks for the PR @sarwarbhuiyan! 
To get the tests to run the `settings.gradle` file needs this added
```
include '_includes:tutorials:schedule-ktable-ttl:kstreams:code'
```
on line 90 right after the previous entry, otherwise, I'm taking a look now

File: _includes/tutorials/schedule-ktable-ttl/kstreams/code/build.gradle [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+    maven {
+        url "https://packages.confluent.io/maven"
+    }
+}
+
+apply plugin: "com.github.johnrengelman.shadow"
+
+dependencies {
+    implementation "org.slf4j:slf4j-simple:1.7.30"
+    implementation "org.apache.kafka:kafka-streams:2.6.0"
```
Comment (bbejeck) [inline _includes/tutorials/schedule-ktable-ttl/kstreams/code/build.gradle:33]: nit: bump the AK version for streams and client and test-utils to `2.7.0`

File: _data/tutorials.yaml [hunk 2]
```diff
@@ -673,3 +673,22 @@ sliding-windows:
     ruby: disabled
     scala: disabled
     swift: disabled
+
+schedule-ktable-ttl:
+  title: "How to implement TTLs with KTable data"
+  meta-description: "How to implement TTLs with KTable data"
+  slug: "/schedule-ktable-ttl"
+  question: "Do you have a requirement to clean up KTable data (in the topic and in the state store) based on TTL"
```
Comment (bbejeck) [inline _data/tutorials.yaml:681]: suggestion - change the `question` section to `How do I implement a TTL clean-up on a KTable?` or something like that.

File: tools/ktablettl.sh [hunk 3]
```diff
@@ -0,0 +1,12 @@
+#!/bin/sh
```
Comment (bbejeck) [inline tools/ktablettl.sh:1]: I don't think we want to commit this

File: _includes/tutorials/schedule-ktable-ttl/kstreams/code/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,5 @@
+FROM confluentinc/cp-kafka-connect-base:5.5.1
```
Comment (bbejeck) [inline _includes/tutorials/schedule-ktable-ttl/kstreams/code/Dockerfile:1]: You can delete this since the tutorial doesn't use datagen or connect

**Review by bbejeck — CHANGES_REQUESTED** [review]
Thanks again for the PR! I've made an initial pass.

**Review by ybyzek — CHANGES_REQUESTED** [review]
Thanks @sarwarbhuiyan !  

https://github.com/confluentinc/kafka-tutorials/pull/752 updated all versions for Confluent, Apache Kafka, Gradle.  Because this PR was started before that PR, it has some versions that also need to be updated -- please update `docker-compose.yml`, `build.gradle`, etc

File: _includes/tutorials/schedule-ktable-ttl/kstreams/code/Dockerfile (hunk 4 above)
Comment (sarwarbhuiyan) [reply _includes/tutorials/schedule-ktable-ttl/kstreams/code/Dockerfile:1]: Deleting the Dockerfile

File: tools/ktablettl.sh (hunk 3 above)
Comment (sarwarbhuiyan) [reply tools/ktablettl.sh:1]: Deleting the script file. I wasn't sure what the use of it was beyond creating the tutorial itself. Will remove.

File: _includes/tutorials/schedule-ktable-ttl/kstreams/markup/test/make-topology-test.adoc [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+      final TestInputTopic<String, String>  inputTopic = testDriver.createInputTopic(inputTopicName,
+                                                                                        keySerializer,
+                                                                                        valueSerializer);
+      final TestInputTopic<String, String>  tableTopic = testDriver.createInputTopic(tableTopicName,
+          keySerializer,
+          valueSerializer);
+
+      final TestOutputTopic<String, String> outputTopic = testDriver.createOutputTopic(outputTopicName, keyDeserializer, valueDeserializer);
+      final TestOutputTopic<String, String> outputTableTopic = testDriver.createOutputTopic(tableTopicName, keyDeserializer, valueDeserializer);
+
```
Comment (bbejeck) [inline _includes/tutorials/schedule-ktable-ttl/kstreams/markup/test/make-topology-test.adoc:50]: Maybe leave this section out as the code below shows how the test is runs and the annotations describe what's going on.

**Review by bbejeck — COMMENTED** [review]
@sarwarbhuiyan, I took another deeper pass and left a few more comments.  Good job on the tutorial!

File: _includes/tutorials/schedule-ktable-ttl/kstreams/markup/dev/make-ttl-emitter.adoc [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+    // application:
+    // * It'll produce the exact same sequence of updates given the same sequence of data.
+    // This seems nice, since the purpose is to modify the data stream itself, you want to
+    // have
+    // a clear understanding of when stuff is going to get deleted. For example, if something
+    // breaks down upstream for this topic, and it stops getting new data for a while, wall
+    // clock
+    // time would just keep deleting data on schedule, whereas stream time will wait for
+    // new updates to come in.
+    // You can change to wall clock time here if that is what is needed
```
[the remark this answers is not in the document]
Comment (sarwarbhuiyan) [reply _includes/tutorials/schedule-ktable-ttl/kstreams/markup/dev/make-ttl-emitter.adoc:44]: Yes, the adoc can be cleaner without the comments. Let me do that.

File: _includes/tutorials/schedule-ktable-ttl/kstreams/markup/dev/make-ttl-emitter.adoc [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+    context.schedule(scanFrequency, PunctuationType.STREAM_TIME, timestamp -> { <2>
+      final long cutoff = timestamp - maxAge.toMillis();
+
+      // scan over all the keys in this partition's store
+      // this can be optimized, but just keeping it simple.
+      // this might take a while, so the Streams timeouts should take this into account
```
[the remark this answers is not in the document]
Comment (sarwarbhuiyan) [reply _includes/tutorials/schedule-ktable-ttl/kstreams/markup/dev/make-ttl-emitter.adoc:50]: By annotations, I guess you mean those <1> things right?

[the remark this answers is not in the document]
Comment (bbejeck) [reply _includes/tutorials/schedule-ktable-ttl/kstreams/markup/dev/make-ttl-emitter.adoc:50]: > By annotations, I guess you mean those <1> things right?

Yes

File: _includes/tutorials/schedule-ktable-ttl/kstreams/markup/test/make-topology-test.adoc (hunk 5 above)
Comment (sarwarbhuiyan) [reply _includes/tutorials/schedule-ktable-ttl/kstreams/markup/test/make-topology-test.adoc]: done

File: _includes/tutorials/schedule-ktable-ttl/kstreams/markup/dev/run-consumer-2.adoc [hunk 8]
```diff
@@ -0,0 +1,24 @@
+////
+  This is a sample content file for how to include a console consumer to the tutorial, probably a good idea so the end user can watch the results
+  of the tutorial.  Change the text as needed.
+
+////
+
+Start a console-consumer to confirm the output (no join happened in the final 3 lines so only the KStream side data should appear in those lines):
+
++++++
+<pre class="snippet"><code class="shell">{% include_raw tutorials/schedule-ktable-ttl/kstreams/code/tutorial-steps/dev/console-consumer-1.sh %}</code></pre>
```
[the remark this answers is not in the document]
Comment (sarwarbhuiyan) [reply _includes/tutorials/schedule-ktable-ttl/kstreams/markup/dev/run-consumer-2.adoc]: nice catch

File: _includes/tutorials/schedule-ktable-ttl/kstreams/markup/dev/make-ttl-emitter.adoc (hunk 6 above)
[the remark this answers is not in the document]
Comment (sarwarbhuiyan) [reply _includes/tutorials/schedule-ktable-ttl/kstreams/markup/dev/make-ttl-emitter.adoc]: done

File: _includes/tutorials/schedule-ktable-ttl/kstreams/markup/dev/make-ttl-emitter.adoc (hunk 7 above)
[the remark this answers is not in the document]
Comment (sarwarbhuiyan) [reply _includes/tutorials/schedule-ktable-ttl/kstreams/markup/dev/make-ttl-emitter.adoc]: done

File: _includes/tutorials/schedule-ktable-ttl/kstreams/markup/dev/make-topology.adoc [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+    // which represents the timestamp
+    final StoreBuilder<KeyValueStore<String, Long>> storeBuilder = Stores.keyValueStoreBuilder( <2>
+        Stores.persistentKeyValueStore(STATE_STORE_NAME),
+        Serdes.String(),
+        Serdes.Long()
+    );
+
+
+    builder.addStateStore(storeBuilder); <3>
+
```
[the remark this answers is not in the document]
Comment (sarwarbhuiyan) [reply _includes/tutorials/schedule-ktable-ttl/kstreams/markup/dev/make-topology.adoc]: done

**Review by bbejeck — APPROVED** [review]
Thanks for the updates @sarwarbhuiyan! LGTM

Just need to confirm @ybyzek's comments are addressed

**Review by ybyzek — APPROVED** [review]
> Just need to confirm @ybyzek's comments are addressed

LGTM.  Thanks for the excellent PR @sarwarbhuiyan 🎉

**Comment by bbejeck** [comment]
The test for this PR passed https://confluentinc.semaphoreci.com/jobs/93c3212e-ab7e-46ad-82c6-dcb5e7d5e605

**Comment by bbejeck** [comment]
The test failure is unrelated to this PR

**Comment by bbejeck** [comment]
Thanks again, @sarwarbhuiyan, for the contribution!

**Comment by sarwarbhuiyan** [comment]
how does it get published to the website?

**Comment by awalther28** [comment]
@sarwarbhuiyan it will get published to the website once we make a release. Do you need it public by any certain date? Right now I think we are waiting on the next ksqlDB release which is in roughly 2 weeks but we can cut a release sooner if need be.

**Pull request merged**